proposals/2023/04-21_build-init-log.md

```diff
@@ -0,0 +1,292 @@
+```
+
+See if we can make the build log sort before the step logs in `ListLogsForBuild`:
+<!-- https://github.com/go-vela/server/blob/main/database/log/list_build.go#L37-L40 -->
+```
+	err = e.client.
+		Table(constants.TableLog).
+		Where("build_id = ?", b.GetID()).
+		Order("step_id ASC").
+```
+
+Sqlite sorts with `NULLS FIRST` by default, and Postgres sorts with `NULLS LAST` (see: [How Are NULLs Sorted by Default?](https://learnsql.com/blog/how-to-order-rows-with-nulls/)).
+So this query is inconsistent between databases. And, the service logs returned by this query are not sorted. It would be nice to have this method return the logs sorted in a consistent manner.
+I would like to see the Build Log, then Step Logs sorted by step_id, then Service Logs sorted by service_id. Any suggestions on how to do that with gorm?
+
+We might need some kind of index or constraint on the `Log` table so that rows with NULL step_id and NULL service_id must have a unique build_id. But, I'm not sure how to create such an index/constraint.
+
+#### Server - API
+
+We will need new endpoints to create/get/update/delete the build init logs (using `/logs` to be consistent with the other endpoints that use `/logs` to get the `Log` associated with a `Step` or `Service`).
+
+```
+- CreateBuildInitLog:  POST   /api/v1/repos/:org/:repo/builds/:build/init/logs
+- GetBuildInitLog:     GET    /api/v1/repos/:org/:repo/builds/:build/init/logs
+- UpdateBuildInitLog:  PUT    /api/v1/repos/:org/:repo/builds/:build/init/logs
+- DeleteBuildInitLog:  DELETE /api/v1/repos/:org/:repo/builds/:build/init/logs
+```
+
+The API path ends up being the only place that needs this "init" string. This avoids ambiguity between retrieve all logs (including step and service logs) and the one `Log` associated only with the `Build` (not a `Step` or `Service`).
+
+These endpoints need to be registered in the router<!-- `LogBuildInitHandlers` in `router/log.go` -->, but no new middleware is required.
+
+Also include the endpoints in the mock server.
+
+#### Server - Compiler
+
+The compiler needs to deprecate and stop adding the `InitStage` and `InitStep`. That includes:
+- remove `compiler/native/initialize{,_test}.go`,
+- remove `Init*` from the `Engine` interface in `compiler/engine.go`,
+- remove the `Init*()` calls from `compiler/native/compile.go`, and
+- remove the magic `"init"` string special-casing in `compiler/native/validate.go`.
+
+Backwards compatibility is a concern with the compiler. Referencing old builds/pipelines should work just fine as it includes the "init" step and stage, thus preserving references to the underlying log. But, recompiling the pipeline will create a slightly different pipeline--one without the injected "init" stage and/or step. The worker will no longer handle the old pipeline with the injected "init" step, so any re-runs MUST re-compile the pipeline.
+
+Does re-running a build ALWAYS re-compile the pipeline?
```

Comment:
Yes, I believe you are correct based off what I'm seeing in the code:

https://github.com/go-vela/server/blob/main/api/build.go#L1200